### ISSUE TITLE
ComboBox: items disappeared

### ISSUE THREAD

**Issue by pinonpierre** [vaadin/framework#11320]
Hi,

I have again a big regression on Vaadin 8.6.0

**The code:**
```
        VerticalLayout mainLayout = new VerticalLayout();
        setCompositionRoot(mainLayout);

        List<String> items = new ArrayList<>(Arrays.asList("blue", "red", "green", "purple", "grey", "orange"));

        ComboBox<String> comboBox = new ComboBox<>();
        comboBox.setPageLength(0);
        comboBox.setItems(items);

        comboBox.setNewItemProvider(value -> {
            items.add(value);
            comboBox.getDataProvider().refreshItem(value); //I don't know if it's required but it doesn't impact the bug
            return Optional.ofNullable(value);
        });
        mainLayout.addComponent(comboBox);
```

**Steps to reproduce the issue:**
- Enter "black" new item in the field.
- Click on the down arrow to shows the items
- 2 items have disappeared: The old "orange" item and the new "black" item

**Small gif to show you the issue:**

![new_item_bug](https://user-images.githubusercontent.com/4507910/48566379-f2eebb80-e8fa-11e8-85dd-7ed0ceeb2cec.gif)

Thanks,

**Comment by TatuLund** [comment]
Could this be related to https://github.com/vaadin/framework/pull/11304 ? And if yes could be fixed in 8.7-SNAPSHOT by https://github.com/vaadin/framework/pull/11304 Needs to be verified.

**Comment by TatuLund** [comment]
I tested this with 8.7-SNAPSHOT and it seems to happen with it also.

However I did further observation. 

- Enter "black" new item in the field.
- Click on the down arrow to shows the items
- 2 items have disappeared: The old "orange" item and the new "black" item
- Enter "black", "orange" or any item that is supposed to be in the list
- Open the drop-down, and you can see all items there

So data provider is updated, but there is momentary glitch

**Comment by ZheSun88** [comment]
need to double check this issue after #11317 got fixed

**Comment by ZheSun88** [comment]
this commit causes this bug https://github.com/vaadin/framework/pull/11094/files, if i remove the new added `private void resetCurrentSuggestionIfNecessary(String selectedKey, String selectedCaption, String selectedIconUri)`, then the behavior will be correct.

**Comment by pinonpierre** [comment]
FYI, in 8.6.2, the behavior is slightly different but still present.

![combobox_bug](https://user-images.githubusercontent.com/4507910/49293720-33306b00-f4b1-11e8-856f-1c993a072eaf.gif)

**Comment by ZheSun88** [comment]
I have come up with two way to solve this issue, but i think they are not good enough, so i want some advice. 
the current `combobox` has different handling on different `PageLength`. when the `pageLength = 0`, the data are not pushed immediately to the client side, instead it is sent when we open the popup (for this issue, the data is sent in 3 steps, first is the `first item`, then the 2nd to the 5th, while there is what we have seen in this ticket. when we open the popup for the second time, the left two items are sent). 

so the two solutions are:

1) we use the same strategy for both `PageLengh` versions, 
   - the problem is that this will increase the size of the initial request;
2) we can request another round of data when open the popup for the first time, which can be achieved   by adding a `beforeClientResponse(true);` to this line https://github.com/vaadin/framework/blob/master/server/src/main/java/com/vaadin/data/provider/DataCommunicator.java#L534 when request the full dataset update. 
 
@Legioth what would be your idea to deal with this bug?

**Comment by pinonpierre** [comment]
Please note that for now, the second time we click on the "down arrow" to unroll the list, the list is unroll and automatically roll after a short time without any action from my part.

![combobox_bug](https://user-images.githubusercontent.com/4507910/49881964-d6796c80-fe2f-11e8-9545-cda3037033f2.gif)

**Comment by TatuLund** [comment]
The original issue described in this ticket seems to be fixed.

What is left here, is known issue https://github.com/vaadin/framework/issues/10915

**Comment by pinonpierre** [comment]
Yes, I confirm that all issues described here are fixed in 8.8.0
